direct/io-elastic/src/main/java/cz/o2/proxima/direct/elastic/ElasticWriter.java

```diff
@@ -0,0 +1,160 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.direct.elastic;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.gson.JsonObject;
+import cz.o2.proxima.direct.core.CommitCallback;
+import cz.o2.proxima.direct.core.OnlineAttributeWriter;
+import cz.o2.proxima.repository.AttributeDescriptor;
+import cz.o2.proxima.storage.StreamElement;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import lombok.extern.slf4j.Slf4j;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.bulk.BulkProcessor;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.xcontent.XContentType;
+
+@Slf4j
+public class ElasticWriter implements OnlineAttributeWriter, BulkProcessor.Listener {
+  private final ElasticAccessor accessor;
+  private final RestHighLevelClient client;
+  private final Map<IndexRequest, CommitCallback> callbacksToCommit = new ConcurrentHashMap<>();
+  private final BulkProcessor bulkProcessor;
+
+  public ElasticWriter(ElasticAccessor accessor) {
+    this.accessor = accessor;
+    this.client = accessor.getRestHighLevelClient();
+    this.bulkProcessor =
+        BulkProcessor.builder(
+                (request, bulkListener) ->
+                    client.bulkAsync(request, RequestOptions.DEFAULT, bulkListener),
+                this)
+            .setBulkActions(accessor.getBatchSize())
+            .setConcurrentRequests(accessor.getConcurrentRequests())
+            .setBulkSize(new ByteSizeValue(10, ByteSizeUnit.MB))
+            .build();
+  }
+
+  @Override
+  public URI getUri() {
+    return accessor.getUri();
+  }
+
+  @Override
+  public void write(StreamElement element, CommitCallback commitCallback) {
+    Preconditions.checkArgument(!element.isDelete(), "Delete not supported.");
+    Preconditions.checkArgument(
+        !element.getAttributeDescriptor().isWildcard(), "Wildcard not supported.");
+
+    final IndexRequest request =
+        new IndexRequest(accessor.getIndexName())
+            .id(element.getKey())
```

Comment:
Looks like the key should be element's key-attribute, no? I'm not familiar with elastic's datamodel, but if documents with the same key are replaced (upserted), then it looks like that.

Reply:
Do you mean wildcard elements? It is true that document will be replaced if the key is the same for wildcard. Maybe this is the real reason why i didn't implemented support for wildcard elements.

Reply:
That is not related to wildcard attributes. The key will be the same for **all** attributes of given entity (key). Seems like in the current version these are overwritten, that is update to attribute X overwrites update to attribute Y.

Reply:
Every attribute has to have own elastic index, otherwise there could be problem with fields mapping. Elastic is not schemaless storage. In older version of elastic there was concept of index types which could be used for different attributes within single index, but it was problematic and removed from elastic. So now is recommended to have single index per data schema.

Reply:
Oh. I think forcing separate index per attribute will not work for me. What is the use-case, that was behind this implementation?

Reply:
I would say that it is best practice. If you share single index for more document types you can have mapping issue. It means that field with the same name always has to have the same field type.  
Or data could be reorganized, instead of actual structure:
```
{"key": "my-key", "attribute": "my-attr", "data": {....}} 
```
we can have: 
{"key": "my-key", "my-attr": {....}}

Reply:
Yes, I understand that inside index documents should have the same structure. But that is fine, I think that we can have the same structure for all attributes. My use-case would be to be able to search for documents (representing entity-key-attribute tuples) with values of some attributes matching a full-text query. I think we can do that by unifying the document and using https://www.elastic.co/guide/en/elasticsearch/plugins/current/mapper-annotated-text-usage.html
We would break down the attributes of the entity and annotate values (as string) with a dot-notation of the inner attribute. 
WDYT?

Reply:
Probably something like 
```
{
"entity": "...",
"key": " ...",
"attribute": "...",
"stamp": "...",
"jsonValue": "value serialized as json string",
"value": "[field_value](field_alias1&field_alias2&...&field_aliasN)
               [field_value](field_alias1&field_alias2&...&field_aliasN)
               ...."
}
```
I think it might be then searchable using the value, and any of the aliases - a field alias would be a stipped dot notation - if we have a nested structure of something like
```proto
message A {
  message B {
    message C {
      int val = 1;
    }
    C c;
  }
  B b;
 }
```
The field would then be `b.c.val`, other aliases would be `c.val` and `val`. That should make possible queries like `value:"field_value=b.c.val"`. Would that work for your case?